tests/test_cmake_build/embedded.cpp

```diff
@@ -0,0 +1,21 @@
+#include <pybind11/embedded.h>
+namespace py = pybind11;
+
+PYBIND11_ADD_EMBEDDED_MODULE(test_cmake_build)(py::module &m) {
+    m.def("add", [](int i, int j) { return i + j; });
```

Comment:
This is one of the things I should have discussed in more detail in the original post. The interface change here is prompted by the need to support the new multi-phase initialization (PEP 489). I haven't actually implemented it yet (it's on the TODO list for this PR) but it requires this interface change. PEP 489 splits up the creation and initialization of modules into separate phases. This makes it possible to support subinterpreters and interpreter reloading. To enable this, it's expected that users only handle module initialization, but not creation.

I think multi-phase initialization support is important enough to warrant the new interface -- at least for embedded modules (although it's certainly possible to have the same interface with extension modules). Note that PEP 489 is only available for Python >= 3.5, but it should be possible backport some of the benefits to older versions of Python by doing a similar creation/initialization split within the macro.

The name change to `MODULE` was mostly done because of the interface change, to avoid having similarly named macros with completely different interfaces. Likewise, the `ADD` in the macro name is intentionally there to suggest that multiple embedded modules can be added to an executable -- as opposed to the single extension module.

The interface inconsistency with `PYBIND11_PLUGIN` is pretty bad and I would normally err on the side of consistency, but I feel that benefits of the new initialization scheme make it worth adopting, not just because of the embedded modules, but also as a way of future-proofing if the old module initialization scheme becomes deprecated within Python. A possible way to restore consistency would be to introduce a similar macro for extension modules:

```C++
PYBIND11_MODULE(name)(py::module &m) { ... } // extension -- only one possible
PYBIND11_ADD_EMBEDDED_MODULE(name)(py::module &m) { ... } // embedded -- any number
```